### DISCUSSION TITLE
Do image transforms have to be regenerated when upgrading to 4.x?

### DISCUSSION THREAD

**Discussion by mikejpeters** [craftcms/cms#13236] [General]
We are in the process of updaing to Craft 4.x and noticed that when viewing pages the first time the images take a long time to load and the image transforms seem to be running again, even though they had been done before.

Is this a necessary part of the upgrade because volumes have changed a bit, or did I make a mistake when upgrading? It seemed like I had to manually create the Amazon S3 filesystems, even though the upgrade docs say "The migration process will take care of volume migrations for you". 

Other than creating the filesystems and making sure the volumes were assigned to the correct filesystems, I didn't make any changes to the volumes or the transforms during the upgrade.

After upgrading, the volume and transform config files are changed as follows:

```diff
diff --git a/config/project/volumes/public--4337f9be-743e-4788-8287-60b76225eac3.yaml 
   75e7ed89-c5d0-4845-858d-2ee6a3696995:
     tabs:
       -
+        elementCondition: null
         elements:
           -
             autocapitalize: true
             autocorrect: true
             class: null
             disabled: false
+            elementCondition: null
             id: null
             instructions: null
             label: null
             step: null
             tip: null
             title: null
-            type: craft\fieldlayoutelements\AssetTitleField
+            type: craft\fieldlayoutelements\assets\AssetTitleField
+            uid: 12e0188b-589e-4e70-ae3b-7bcf3e6e5402
+            userCondition: null
             warning: null
             width: 100
         name: Content
-        sortOrder: 1
+        uid: 62d09118-8c7e-48e7-acfc-de99cd9e6343
+        userCondition: null
+fs: public
 handle: public
-hasUrls: true
 name: Public
-settings:
-  addSubfolderToRootUrl: '1'
-  autoFocalPoint: ''
-  bucket: ***
-  bucketSelectionMode: manual
-  cfDistributionId: ''
-  cfPrefix: ''
-  expires: ''
-  keyId: ***
-  makeUploadsPublic: '1'
-  region: ca-central-1
-  secret: ***
-  storageClass: ''
-  subfolder: ''
 sortOrder: 3
 titleTranslationKeyFormat: null
 titleTranslationMethod: site
-type: craft\awss3\Volume
-url: 'https://s3.ca-central-1.amazonaws.com/***/'
+transformFs: ''
+transformSubpath: ''
```

```diff
diff --git a/config/project/imageTransforms/animalProfile--e4fd11cb-7772-4a6e-8bc8-1ba09ca830ad.yaml 
+fill: null
 format: jpg
 handle: animalProfile
 height: 400
 name: 'Animal Profile'
 position: center-center
 quality: 60
+upscale: true
 width: 400
```

Thanks for any help you can provide!

**Comment by brianjhanson** [comment]
Transforms shouldn't need to be regenerated in this scenario. You might be right that something deviated in your upgrade. Especially since you mentioned having to re-create the S3 filesystem. That should also be handled in the migration. 

That said, in testing this, I noticed if you have `upscaleImages` set to `false` and are defining your transforms in your templates, you will see those images regenerated. When `upscaleImages` is set to `false`, the transform folder names will get appended with `_ns` (no-scale). That will result in those images getting regenerated as the folder path has changed. We're going to chat internally to see if there's a good way to smooth that over. 

That doesn't seem to be what's happening here, though. You're using named transforms (so the transform folder name remains the same), and you have `upscale` set to `true` (which doesn't append anything to the folder names). I'm only mentioning it on the off chance you have some transforms defined in your templates and have changed your global `upscaleImages` setting to `false` somewhere along the way.

I'm going to close this out, but if you have additional details or questions I'm happy to help out more.

**Comment by brandonkelly** [comment]
>  It seemed like I had to manually create the Amazon S3 filesystems, even though the upgrade docs say "The migration process will take care of volume migrations for you".

We’ve seen this happen if the initial upgrade fails, and then is reattempted without restoring the site to its pre-upgrade state first.

When a migration fails, you should:

1. Restore a database backup from before the update.
2. Revert any changes that the update attempt made to your `config/project/` folder.
